src/udev/udev-ctrl.c

```diff
@@ -205,12 +205,12 @@ static int udev_ctrl_connection_event_handler(sd_event_source *s, int fd, uint32
                 return 0;
         }
 
-        if (msg_wire.type == _UDEV_CTRL_END_MESSAGES)
-                return 0;
-
         if (uctrl->callback)
                 (void) uctrl->callback(uctrl, msg_wire.type, &msg_wire.value, uctrl->userdata);
 
+        if (msg_wire.type == _UDEV_CTRL_END_MESSAGES)
```

Comment:
it is not clear to me how shuffling around these and the additional callback help with the problem, could you please add more details to the commit message?